api/src/main/java/net/kyori/adventure/resourcepack/ResourcePackImpl.java

```diff
@@ -0,0 +1,98 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.resourcepack;
+
+import net.kyori.adventure.text.Component;
+import net.kyori.examination.ExaminableProperty;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
+import java.util.stream.Stream;
+
+import static java.util.Objects.requireNonNull;
+
+final class ResourcePackImpl implements ResourcePack {
+  private final String url;
+  private final String hash;
+  private final boolean required;
+  private final Component prompt;
+
+  ResourcePackImpl(final @NotNull String url, final @NotNull String hash, final boolean required, final @Nullable Component prompt) {
+    this.url = requireNonNull(url, "url");
+    this.hash = requireNonNull(hash, "hash");
+    this.required = required;
+    this.prompt = prompt;
+  }
+
+  @Override
+  public @NotNull String url() {
+    return this.url;
+  }
+
+  @Override
+  public @NotNull String hash() {
+    return this.hash;
+  }
+
+  @Override
+  public boolean required() {
+    return this.required;
+  }
+
+  @Override
+  public @Nullable Component prompt() {
+    return this.prompt;
+  }
+
+  @Override
+  public @NotNull Stream<? extends ExaminableProperty> examinableProperties() {
+    return Stream.of(
+      ExaminableProperty.of("url", this.url),
+      ExaminableProperty.of("hash", this.hash),
+      ExaminableProperty.of("required", this.required),
+      ExaminableProperty.of("prompt", this.prompt)
+    );
+  }
+
+  @Override
+  public boolean equals(final Object o) {
+    if (this == o) return true;
+    if (!(o instanceof ResourcePackImpl)) return false;
+    final ResourcePackImpl that = (ResourcePackImpl) o;
+    return this.required == that.required
+      && this.url.equals(that.url)
+      && this.hash.equals(that.hash)
+      && Objects.equals(this.prompt, that.prompt);
+  }
+
+  @Override
+  public int hashCode() {
+    int result = this.url.hashCode();
+    result = 31 * result + this.hash.hashCode();
+    result = 31 * result + (this.required ? 1 : 0);
+    result = 31 * result + (this.prompt != null ? this.prompt.hashCode() : 0);
+    return result;
+  }

```

Comment:
make sure to override toString() to call out to `Internals.toString()` -- this computes it based on the Examinable information.